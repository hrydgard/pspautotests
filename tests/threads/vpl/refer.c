#include "shared.h"

void testRefer(const char *title, SceUID vpl, SceKernelVplInfo *info) {
	int result = sceKernelReferVplStatus(vpl, info);
	if (result == 0) {
		schedf("%s: ", title);
		schedVplInfo(info);
	} else {
		schedf("%s: Failed (%X)\n", title, result);
	}
}

int main(int argc, char **argv) {
	SceUID vpl = sceKernelCreateVpl("refer", PSP_MEMORY_PARTITION_USER, 0, 0x10000, NULL);
	SceKernelVplInfo info;
	info.size = sizeof(info);

	// Crashes.
	//testRefer("NULL info", vpl, NULL);
	testRefer("Normal", vpl, &info);

	schedf("\nSizes:\n");
	int sizes[] = {0, 1, 2, 4, 8, 44, 48, 52, 56, 64, 128, 256, 1024, 0xFF00, 0xFF0000, 0x80000001, 0xCCCCCCCC, 0xFFFFFFFF};
	int i, result;
	for (i = 0; i < sizeof(sizes) / sizeof(sizes[0]); ++i) {
		info.size = sizes[i];
		result = sceKernelReferVplStatus(vpl, &info);
		schedf("  %08X => %08X (result=%08X)\n", sizes[i], info.size, result);
	}
	schedf("\n");
	flushschedf();

	sceKernelDeleteVpl(vpl);

	testRefer("NULL", 0, &info);
	testRefer("Invalid", 0xDEADBEEF, &info);
	testRefer("Deleted", vpl, &info);

	BASIC_SCHED_TEST("NULL",
		result = sceKernelReferVplStatus(0, &info);
	);
	BASIC_SCHED_TEST("Refer other",
		result = sceKernelReferVplStatus(vpl2, &info);
	);
	BASIC_SCHED_TEST("Refer same",
		result = sceKernelReferVplStatus(vpl1, &info);
	);

	flushschedf();
	return 0;
}