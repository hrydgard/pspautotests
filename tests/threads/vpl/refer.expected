Normal: VPL: OK (size=52,name=refer,attr=00000000,poolSize=0000FFE0,freeSize=0000FFE0,wait=0)

Sizes:
  00000000 => 00000000 (result=00000000)
  00000001 => 00000034 (result=00000000)
  00000002 => 00000034 (result=00000000)
  00000004 => 00000034 (result=00000000)
  00000008 => 00000034 (result=00000000)
  0000002C => 00000034 (result=00000000)
  00000030 => 00000034 (result=00000000)
  00000034 => 00000034 (result=00000000)
  00000038 => 00000034 (result=00000000)
  00000040 => 00000034 (result=00000000)
  00000080 => 00000034 (result=00000000)
  00000100 => 00000034 (result=00000000)
  00000400 => 00000034 (result=00000000)
  0000FF00 => 00000034 (result=00000000)
  00FF0000 => 00000034 (result=00000000)
  80000001 => 00000034 (result=00000000)
  CCCCCCCC => 00000034 (result=00000000)
  FFFFFFFF => 00000034 (result=00000000)

NULL: Failed (8002019C)
Invalid: Failed (8002019C)
Deleted: Failed (8002019C)
NULL: A1B1C2E1D1F2 (thread=800201B5, main=8002019C)
Refer other: A1B1C2E1D1F2 (thread=800201B5, main=00000000)
Refer same: A1B1C2E1D1F2 (thread=800201B5, main=00000000)
